### PR TITLE
Derived fields part 1: support counts

### PR DESCRIPTION
**Pull request by viswajithiii**
## Description

Support ordering by counts, and refactor derived fields so it's not as manual as it was with DackBox.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Unit tests

### CODE REVIEW

File: pkg/postgres/walker/walker.go [hunk 1]
```diff
@@ -221,18 +222,29 @@ func getProtoBufName(protoBufTag string) string {
...
 
-func getSearchOptions(ctx context, searchTag string) SearchField {
+func getSearchOptions(ctx context, searchTag string) (SearchField, []DerivedSearchField) {
 	ignored := searchTag == "-"
 	if ignored || searchTag == "" {
 		return SearchField{
 			Ignored: ignored,
+		}, nil
+	}
+	field, _ := stringutils.Split2(searchTag, ",")
```
Comment (connorgorman) [inline pkg/postgres/walker/walker.go:232]: stringutils.GetUpTo(",")

File: pkg/postgres/walker/walker.go [hunk 2]
```diff
@@ -221,18 +222,29 @@ func getProtoBufName(protoBufTag string) string {
...
+	derivedSearchFieldsMap := search.GetFieldsDerivedFrom(field)
+	var derivedSearchFields []DerivedSearchField
```
Comment (connorgorman) [inline pkg/postgres/walker/walker.go:234]: Mega nit: switch these two lines

File: pkg/search/options.go [hunk 3]
```diff
@@ -313,16 +309,78 @@ var (
...
+	return ok
+}
+
+// GetFieldsDerivedFrom gets the fields derived from the given search field.
+func GetFieldsDerivedFrom(s string) map[string]DerivationType {
+	return derivationsByField[strings.ToLower(s)]
+}
+
+func newFieldLabelWithMetadata(s string, metadata *DerivedFieldLabelMetadata) FieldLabel {
+	lowerS := strings.ToLower(s)
```
Comment (connorgorman) [inline pkg/search/options.go:351]: for some reason lowerS is strange to me. just set 
`s = strings.ToLower(s)`

File: pkg/search/options.go [hunk 4]
```diff
@@ -313,16 +309,78 @@ var (
...
+// DerivedFieldLabelMetadata includes metadata showing that a field is derived.
+type DerivedFieldLabelMetadata struct {
+	DerivedFrom    FieldLabel
+	DerivationType DerivationType
+}
+
+// DerivationType represents a type of derivation.
+type DerivationType int
+
+// THis block enumerates all supported derivation types.
```
Comment (connorgorman) [inline pkg/search/options.go:383]: THis -> This

File: pkg/search/postgres/common.go [hunk 5]
```diff
@@ -147,6 +148,14 @@ func (q *query) AsSQL() string {
 		querySB.WriteString(" where ")
 		querySB.WriteString(replaceVars(q.Where))
 	}
+	if q.GroupByPrimaryKey {
+		var primaryKeyPaths []string
```
Comment (connorgorman) [inline pkg/search/postgres/common.go:152]: presize

File: pkg/search/options.go (hunk 3 above)
Comment (viswajithiii) [reply pkg/search/options.go:351]: I can't, unfortunately, because I want to return `FieldLabel(s)` below.

File: pkg/search/postgres/joins.go [hunk 6]
```diff
@@ -132,7 +137,17 @@ func getJoinsAndFields(src *walker.Schema, q *v1.Query) ([]innerJoin, map[string
...
 			if unreachedFields.Remove(lowerCaseName) {
-				reachableFields[lowerCaseName] = &field
+				reachableFields[lowerCaseName] = searchFieldMetadata{baseField: &field}
+			}
+			for _, derivedF := range field.DerivedSearchFields {
+				derivedField := derivedF
+				lowerCaseDerivedName := strings.ToLower(derivedField.FieldName)
+				if unreachedFields.Remove(lowerCaseDerivedName) {
+					reachableFields[lowerCaseDerivedName] = searchFieldMetadata{
+						baseField:       &field,
```
Comment (md2119) [inline pkg/search/postgres/joins.go:147]: Why do we add the base field with derived field?

Comment (viswajithiii) [reply pkg/search/postgres/joins.go:147]: Because you need the base field for constructing the query, since that's the column in the DB

Comment (md2119) [reply pkg/search/postgres/joins.go:147]: oh okay. Thanks for clarifying. I thought we already know `DerivedFrom` and `DerivationType`.

Comment (viswajithiii) [reply pkg/search/postgres/joins.go:147]: We do, this is just transferring that information

**Pull request merged**